driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java

```diff
@@ -187,445 +144,603 @@ private void logError(InetSocketAddress address, Throwable exception) {
-                            write(connection, prepareAndRetry(toPrepare.getQueryString()));
-                            // we're done for now, the prepareAndRetry callback will handle the rest
-                            return;
-                        default:
-                            if (metricsEnabled())
-                                metrics().getErrorMetrics().getOthers().inc();
-                            break;
-                    }
+        void sendRequest() {
+            try {
+                while (!isDone.get() && queryPlan.hasNext() && !queryStateRef.get().isCancelled()) {
+                    Host host = queryPlan.next();
+                    logger.trace("[{}] Querying node {}", id, host);
+                    if (query(host))
+                        return;
+                }
+                reportNoMoreHosts(this);
+            } catch (Exception e) {
+                // Shouldn't happen really, but if ever the loadbalancing policy returned iterator throws, we don't want to block.
+                setFinalException(null, new DriverInternalError("An unexpected error happened while sending requests", e));
+            }
+        }
 
-                    if (retry == null)
-                        setFinalResult(connection, response);
-                    else {
-                        switch (retry.getType()) {
-                            case RETRY:
-                                ++retriesByPolicy;
-                                if (logger.isDebugEnabled())
-                                    logger.debug("Doing retry {} for query {} at consistency {}", retriesByPolicy, statement, retry.getRetryConsistencyLevel());
-                                if (metricsEnabled())
-                                    metrics().getErrorMetrics().getRetries().inc();
-                                retry(true, retry.getRetryConsistencyLevel());
-                                break;
-                            case RETHROW:
-                                setFinalResult(connection, response);
-                                break;
-                            case IGNORE:
-                                if (metricsEnabled())
-                                    metrics().getErrorMetrics().getIgnores().inc();
-                                setFinalResult(connection, new Responses.Result.Void());
-                                break;
-                        }
-                    }
-                    break;
-                default:
-                    setFinalResult(connection, response);
-                    break;
+        private boolean query(final Host host) {
+            currentPool = manager.pools.get(host);
+            if (currentPool == null || currentPool.isClosed())
+                return false;
+
+            if (nextExecutionScheduled.compareAndSet(false, true))
+                scheduleExecution(speculativeExecutionPlan.nextExecution(host));
```

Comment:
Do we plan on adding some kind of control to SpeculativeExecutionPolicy like we do for RetryPolicy with RetryDecision?  My first expectation was that speculative retries would only be for reads, but I can understand why you would want to do it for mutations to in some cases.

I am thinking for the same reason we have onWrite/onReadTimeout in RetryPolicy that we'll want to control whether or not we schedule another execution.  Some scenarios I can think of:

1. counter updates (if a speculative retry gets executed and the first execution does eventually succeed, you'll increment a counter twice).
2. mutations in general.  i.e. if an update gets speculative retried and succeeds, user gets the confirmation and performs another update (for some reason), and then the first update from the first query succeeds (maybe it was slowed down by network latency on a connection).  If the user understands this could happen it may not be a big deal.